File: verifier_fabric/src/main/java/eu/toldi/zokratescontract/ZoKrates.java
/*
 * Copyright 2023 Contributors of the zkWF project
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package eu.toldi.zokratescontract;

import eu.toldi.zokratescontract.exceptions.ZokratesInstallFailed;
import eu.toldi.zokratescontract.exceptions.ZokratesNotInititialised;
import eu.toldi.zokratescontract.exceptions.ZokratesVerificationFailed;
import org.hyperledger.fabric.Logger;

import java.io.*;
import java.net.URL;
import java.util.Objects;

public enum ZoKrates {
    INSTANCE;
    private boolean isInitDone = false;
    private final Logger logger = Logger.getLogger("Zokrates");
    private String zokratesPath;


    private void checkInit() throws ZokratesNotInititialised {
        if (!isInitDone) {
            throw new ZokratesNotInititialised();
        }
    }

    private void installZokrates() throws IOException, InterruptedException, ZokratesInstallFailed {
        File tmp = new File(System.getProperty("user.dir") + "/tmp");
        tmp.mkdir();
        File dlFile = new File(tmp,"zokrates.sh");
        logger.info(dlFile.getAbsoluteFile().getAbsolutePath().toString());
        try (BufferedInputStream inputStream = new BufferedInputStream(new URL("https://raw.githubusercontent.com/ZoKrates/ZoKrates/master/scripts/one_liner.sh").openStream());
             FileOutputStream fileOS = new FileOutputStream(dlFile)) {
            byte data[] = new byte[1024];
            int byteContent;
            while ((byteContent = inputStream.read(data, 0, 1024)) != -1) {
                fileOS.write(data, 0, byteContent);
            }
        } catch (IOException e) {
            // handles IO exceptions
        }
        dlFile.setExecutable(true);
        ProcessBuilder pb = new ProcessBuilder(dlFile.getAbsoluteFile().getAbsolutePath().toString(), "--to", ".zokrates");
        Process process = pb.start();
        int exitCode = process.waitFor();
        if (exitCode != 0) {
            throw new ZokratesInstallFailed();
        }
        dlFile.delete();
        tmp.delete();
        logger.info("ZoKrates installed");
    }

    public void initZokrates() throws IOException, InterruptedException, ZokratesInstallFailed {
        if ( !(new File(System.getProperty("user.dir") + "/.zokrates/bin/zokrates").exists())) {
            installZokrates();
        }
        zokratesPath = System.getProperty("user.dir") + "/.zokrates/bin/zokrates";
        isInitDone = true;
        logger.info("ZoKrates initialized");
    }

    public boolean verify(zkProof p, VerificationKey v) throws ZokratesNotInititialised, IOException, InterruptedException {
        checkInit();
        File verificationKey = new File(System.getProperty("user.dir") +"/verification.key");
        File proofJson = new File(System.getProperty("user.dir") +"/proof.json");
        writeToFile(verificationKey,v.toJSONString());
        writeToFile(proofJson,p.toJSONString());

        ProcessBuilder pb = new ProcessBuilder(zokratesPath, "verify");
        Process process = pb.start();
        int exitCode = process.waitFor();
        String result = null;
        if (exitCode == 0) {
            BufferedReader reader =
                    new BufferedReader(new InputStreamReader(process.getInputStream()));
            StringBuilder builder = new StringBuilder();
            String line = null;
            while ( (line = reader.readLine()) != null) {
                builder.append(line);
                builder.append(System.getProperty("line.separator"));
            }
            result = builder.toString();

        }else {
            throw new ZokratesVerificationFailed();
        }

        proofJson.delete();
        verificationKey.delete();
        return Objects.equals(result.split("\n")[1], "PASSED");
    }

    private void writeToFile(File file,String data){
        try(FileOutputStream fos = new FileOutputStream(file);
            BufferedOutputStream bos = new BufferedOutputStream(fos)) {
            //convert string to byte array
            byte[] bytes = data.getBytes();
            //write byte array to file
            bos.write(bytes);
            bos.close();
            fos.close();
            System.out.print("Data written to file successfully.");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: generator/zokrates-wrapper/src/main/kotlin/eu/toldi/zokrates-wrapper/Zokrates.kt
/*
 * Copyright 2023 Contributors of the zkWF project
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package eu.toldi.`zokrates-wrapper`

import eu.toldi.`zokrates-wrapper`.exceptions.*
import java.io.File
import java.net.URL

object Zokrates {
    private var isInitDone = false
    private lateinit var zokratesPath: String

    private fun checkInit() {
        if (isInitDone.not()) {
            throw ZokratesNotInititialised()
        }
    }

    private fun installZokrates() {
        val tmp = File(System.getProperty("user.dir") + "/tmp").also { file ->
            file.mkdir()
        }
        val dlFile = File(tmp, "zokrates.sh").also {
            it.writeText(URL("https://raw.githubusercontent.com/ZoKrates/ZoKrates/master/scripts/one_liner.sh").readText())
            it.setExecutable(true)
        }

        val pb = ProcessBuilder(dlFile.absoluteFile.absolutePath.toString(), "--to", ".zokrates")
        val process = pb.start()

        val exitCode = process.waitFor()
        if (exitCode != 0) {
            throw ZokratesInstallFailed()
        }

        dlFile.delete()
        tmp.delete()
    }

    fun initZokrates() {
        if (File(System.getProperty("user.dir") + "/.zokrates/bin/zokrates").exists().not()) {
            installZokrates()
        }

        zokratesPath = System.getProperty("user.dir") + "/.zokrates/bin/zokrates"
        isInitDone = true
    }

    fun compile(path: String, outPut: String = "out") {
        checkInit()
        compile(File(path), outPut)
    }

    fun compile(file: File, outPut: String = "out") {
        checkInit()
        val pb = ProcessBuilder(zokratesPath, "compile", "-i", file.absoluteFile.absolutePath.toString(), "-o", outPut)

        val process = pb.start()

        val exitCode = process.waitFor()
        if (exitCode != 0) {
            val output = process.inputStream.bufferedReader().use { it.readText() }
            throw ZokratesCompileFailed(output)
        }
    }

    fun computeWithness(args: List<String>, input: String = "out", output: String = "witness"): List<String> {
        checkInit()
        val sb = StringBuilder()
        var prefix = ""
        args.forEach {
            sb.append(prefix)
            prefix = " "
            sb.append(it)
        }
        println(sb.toString())
        val command = mutableListOf(zokratesPath, "compute-witness", "-i", input, "-o", output, "-a").apply {
            addAll(args)
        }
        val processBuilder = ProcessBuilder(command)

        val process = processBuilder.start()

        val exitCode = process.waitFor()
        if (exitCode != 0) {
            val output1 = process.inputStream.bufferedReader().use { it.readText() }
            println("Error $output1")
            throw ZokratesComputeWitnessFailed(output1)
        }
        return readWitness(output)
    }

    fun readWitness(witness: String = "witness"): List<String> {
        return File(witness).readText().split("\n").filter { it.startsWith("~out_") }.sorted().map { it.split(' ')[1] }
    }

    fun setup(input: String = "out") {
        val pb = ProcessBuilder(zokratesPath, "setup", "-i", input)
        val process = pb.start()
        val exitCode = process.waitFor()
        if (exitCode != 0) {
            val output = process.inputStream.bufferedReader().use { it.readText() }
            println("Error $output")
            throw ZokratesSetupFailed(output)
        }
    }

    fun generateProof(input: String = "out", witness: String = "witness", proof: String = "proof.json") {
        val pb = ProcessBuilder(zokratesPath, "generate-proof", "-i", input, "-w", witness, "-j", proof)
        val process = pb.start()
        val exitCode = process.waitFor()
        if (exitCode != 0) {
            val output = process.inputStream.bufferedReader().use { it.readText() }
            println("Error $output")
            throw ZokratesSetupFailed(output)
        }
    }

    fun exportVerifier(input: String = "out", output: String = "verifier.sol") {
        val pb = ProcessBuilder(zokratesPath, "export-verifier")
        val process = pb.start()
        val exitCode = process.waitFor()
        if (exitCode != 0) {
            val output = process.inputStream.bufferedReader().use { it.readText() }
            println("Error $output")
            throw ZokratesSetupFailed(output)
        }
    }
}